spec.emu

```diff
@@ -997,23 +1160,22 @@ contributors: Google, Ecma International
           It may throw a *RangeError* exception if _overflow_ is ~reject~ and the resulting month or day would need to be clamped in order to form a valid date in _calendar_.
         </dd>
       </dl>
-      <p>The behaviour is implementation-defined, but all calendars follow the general steps given here, which is a generalization of the precise algorithm specified in CalendarDateAdd for *"iso8601"*.</p>
+      <p>All calendars follow the steps given here, which is a generalization of the precise algorithm specified in CalendarDateAdd for *"iso8601"*.</p>
       <p>This definition supersedes the definition provided in <emu-xref href="#sec-temporal-nonisodateadd"></emu-xref>.</p>
       <p>It performs the following steps when called:</p>
       <emu-alg>
-        1. Let _calendarDate_ be CalendarISOToDate(_calendar_, _isoDate_).
-        1. Add _duration_.[[Years]] to _calendarDate_ without changing any less-significant fields.
-        1. If YearContainsMonthCode(_calendar_, _calendarDate_.[[Year]], _calendarDate_.[[MonthCode]]) is *false*, then
-          1. NOTE: This only happens in lunisolar calendars.
-          1. Set _calendarDate_.[[MonthCode]] to ! ConstrainMonthCode(_calendar_, _calendarDate_.[[Year]], _calendarDate_.[[MonthCode]], ~constrain~).
-          1. Set _calendarDate_.[[Month]] to MonthCodeToOrdinal(_calendar_, _calendarDate_.[[Year]], _calendarDate_.[[MonthCode]]).
-        1. Add _duration_.[[Months]] to _calendarDate_, balancing _calendarDate_ if it goes over a year boundary.
-        1. If the date described by _calendarDate_ does not exist, then
+        1. Let _parts_ be CalendarISOToDate(_calendar_, _isoDate_).
+        1. Let _y0_ be _parts_.[[Year]] + _duration_.[[Years]].
+        1. Let _m0_ be MonthCodeToOrdinal(_calendar_, _y0_, ! ConstrainMonthCode(_calendar_, _y0_, _parts_.[[MonthCode]], ~constrain~)).
+        1. Let _endOfMonth_ be BalanceNonISODate(_calendar_, _y0_, _m0_ + _duration_.[[Months]] + 1, 0).
+        1. Let _baseDay_ be _parts_.[[Day]].
+        1. If _endOfMonth_.[[Day]] &lt; _baseDay_, then
```

Comment:
Internal consistency nit:
```suggestion
        1. If _baseDay_ > _endOfMonth_.[[Day]], then
```
or probably better still:
```suggestion
        1. If _baseDay_ &lt; _endOfMonth_.[[Day]], then
          1. Let _regulatedDay_ be _baseDay_.
        1. Else,
          1. If _overflow_ is ~reject~, throw a *RangeError* exception.
          1. Let _regulatedDay_ be _endOfMonth_.[[Day]].
```
(and likewise in NonISOCalendarDateToISO)